pkg/disruption/ci/factory.go

```diff
@@ -0,0 +1,238 @@
+
+const (
+	ProtocolHTTP1 ProtocolType = "http1"
+	ProtocolHTTP2 ProtocolType = "http2"
+)
+
+type LoadBalancerType string
+
+const (
+	ExternalLoadBalancerType LoadBalancerType = "external-lb"
+	InternalLoadBalancerType LoadBalancerType = "internal-lb"
+	ServiceNetworkType       LoadBalancerType = "service-network"
+)
+
+type ServerNameType string
+
+const (
+	KubeAPIServer      ServerNameType = "kube-apiserver"
+	OpenShiftAPIServer ServerNameType = "openshift-apiserver"
+)
+
+const (
+	KubeAPIServerShutdownIntervalName    = "openshift-kube-apiserver-shutdown"
+	KubeAPIServerShutdownIntervalLocator = "disruption/graceful-shutdown server/kube-apiserver"
+)
+
+// Factory creates a new instance of a Disruption test from
+// the user specified configuration.
+type Factory interface {
+	New(TestConfiguration) (*BackendSampler, error)
+}
+
+// NewDisruptionTestFactory returns a shared disruption test factory that uses
+// the given rest Config object to create new disruption test instances.
+func NewDisruptionTestFactory(config *rest.Config) Factory {
+	return &testFactory{
+		dependency: &restConfigDependency{
+			config: config,
+		},
+	}
+}
+
+// TestConfiguration allows a user to specify the disruption test parameters
+type TestConfiguration struct {
+	TestType
+
+	// Path is the request path that the backend sampler will exercise
+	Path string
+
+	// Timeout is the transport timeout, it is the maximum amount of time a
+	// dial will wait for a connect to complete.
+	// If Deadline is also set, it may fail earlier.
+	// NOTE: we infer the client timeout and request deadline from this value,
+	//  this is how the current backend sampler works.
+	Timeout time.Duration
+
+	// SampleInterval is the interval that the sampler will
+	// wait before generating the next sample.
+	// NOTE: the current backend sampler implementation defaults
+	//  it to 1s.
+	SampleInterval time.Duration
+
+	// EnableShutdownResponseHeader indicates whether to include the shutdown
+	// response header extractor, this should be true only when the
+	// request(s) are being sent to the kube-apiserver.
+	EnableShutdownResponseHeader bool
+}
+
+// TestType defines the disruption test type, the user must
+// provide a complete specification for the desired test.
+type TestType struct {
+	// TargetServer is the server that is is being tested
+	TargetServer ServerNameType
+
+	// LoadBalancerType is the type of load balancer through which the
+	// disruption test is hitting the target server.
+	LoadBalancerType LoadBalancerType
+
+	// ConnectionType specifies whether the underlying TCP connection(s) used
+	// by the requests should be new or reused.
+	ConnectionType monitorapi.BackendConnectionType
+
+	// Protocol specifies the protocol used by the test,
+	// whether it is http/1x or http/2.0
+	Protocol ProtocolType
+}
+
+func (t TestType) Name() string {
+	return fmt.Sprintf("backend-sampler-%s-%s-%s-%s", t.TargetServer, t.Protocol, t.LoadBalancerType, t.ConnectionType)
```

Comment:
Regarding BackendName, I think it may not be used, so it's not critical, but it would be nice to be consistent with the existing entries just in case something is using it. In our case, I think the BackendName should just be Name, minus the [new|reused]-connections suffix.

So:
Name: backend-sampler-openshift-apiserver-http2-external-lb-reused-connections
BackendName: backend-sampler-openshift-apiserver-http2-external-lb